internal/redirecttraffic/redirect_traffic.go

```diff
@@ -0,0 +1,205 @@
+		EnableConsulDNS:      cfg.ConsulDNSEnabled(),
+		ExcludeInboundPorts:  cfg.TransparentProxy.ExcludeInboundPorts,
+		ExcludeOutboundPorts: cfg.TransparentProxy.ExcludeOutboundPorts,
+		ExcludeOutboundCIDRs: cfg.TransparentProxy.ExcludeOutboundCIDRs,
+		ExcludeUIDs:          cfg.TransparentProxy.ExcludeUIDs,
+	}
+
+	for _, opt := range opts {
+		opt(input)
+	}
+
+	return input
+}
+
+// applyTrafficRedirectionRules creates and applies traffic redirection rules with
+// the help of iptables
+//
+// iptables.Config:
+//
+//	ConsulDNSIP: Consul Dataplane's DNS server (i.e. localhost)
+//	ConsulDNSPort: Consul Dataplane's DNS server's bind port
+//	ProxyUserID: a constant set by default in the mesh-task module for the Consul dataplane's container
+//	ProxyInboundPort: the default inbound port or bind port
+//	ProxyOutboundPort: default transparent proxy outbound port
+//	ExcludeInboundPorts: prometheus, envoy stats, expose paths and `transparentProxy.excludeInboundPorts`
+//	ExcludeOutboundPorts: `transparentProxy.excludeOutboundPorts` in CONSUL_ECS_CONFIG_JSON
+//	ExcludeOutboundCIDRs: TaskMeta IP, Consul Server IP and `transparentProxy.excludeOutboundCIDRs` in CONSUL_ECS_CONFIG_JSON
+//	ExcludeUIDs: `transparentProxy.excludeUIDs` in CONSUL_ECS_CONFIG_JSON
+func (c *TrafficRedirectionCfg) Apply() error {
+	if c.ProxySvc == nil {
+		return fmt.Errorf("proxy service details are required to enable traffic redirection")
+	}
+
+	// Decode proxy's opaque config
+	var trCfg trafficRedirectProxyConfig
+	if err := mapstructure.WeakDecode(c.ProxySvc.Proxy.Config, &trCfg); err != nil {
+		return fmt.Errorf("failed parsing proxy service's Proxy.Config: %w", err)
+	}
+
+	c.iptablesCfg = iptables.Config{
+		ProxyUserID:       strconv.Itoa(defaultProxyUserID),
+		ProxyInboundPort:  defaultProxyInboundPort,
+		ProxyOutboundPort: iptables.DefaultTProxyOutboundPort,
+	}
+
+	// Override proxyInboundPort with bind_port
+	if trCfg.BindPort != 0 {
+		c.iptablesCfg.ProxyInboundPort = trCfg.BindPort
+	}
+
+	// Override the outbound port if the outbound port present in the proxy registration
+	if c.ProxySvc.Proxy.TransparentProxy != nil && c.ProxySvc.Proxy.TransparentProxy.OutboundListenerPort != 0 {
+		c.iptablesCfg.ProxyOutboundPort = c.ProxySvc.Proxy.TransparentProxy.OutboundListenerPort
+	}
+
+	// Inbound ports
+	{
+		for _, port := range c.ExcludeInboundPorts {
+			c.iptablesCfg.ExcludeInboundPorts = append(c.iptablesCfg.ExcludeInboundPorts, strconv.Itoa(port))
+		}
+
+		// Exclude Envoy's ready bind port that indicate envoy's readiness
+		c.iptablesCfg.ExcludeInboundPorts = append(c.iptablesCfg.ExcludeInboundPorts, strconv.Itoa(c.ProxyHealthCheckPort))
+
+		// Exclude envoy_prometheus_bind_addr port from inbound redirection rules.
+		if trCfg.PrometheusBindAddr != "" {
+			_, port, err := net.SplitHostPort(trCfg.PrometheusBindAddr)
+			if err != nil {
+				return fmt.Errorf("failed parsing host and port from envoy_prometheus_bind_addr: %w", err)
```

Comment:
should we log `PrometheusBindAddr` or it will be logged by itself?

Reply:
Address gets logged automatically in the `err` variable.